### PR TITLE
docs(specs): Customizability

### PR DESCRIPTION
**Pull request by fridrik01**
Added documentation for IPC Customazbility

### CODE REVIEW

File: specs/customizability.md [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+#[delegate(SendOps<K>, generics = "K", where = "K: CustomKernel")]
+#[delegate(UpgradeOps<K>, generics = "K", where = "K: CustomKernel")]
+pub struct CustomKernelImpl<C>(pub DefaultKernel<C>);
+```
+
+[fendermint/vm/interpreter/src/fvm/examples/mycustomkernel.rs#L27](https://github.com/consensus-shipyard/ipc/blob/98497363a10e08236325e6d5c52755b9fcd52958/fendermint/vm/interpreter/src/fvm/examples/mycustomkernel.rs#L27)
+
+### **2. Implementing all necessary functions for the syscall**
+
+Implement `my_custom_syscall` Here is where we implement our custom syscall:
```
Comment (folex) [inline specs/customizability.md:77]: > plement `my_custom_syscall` Here is

newline or a header missed before `Here`?

File: specs/customizability.md [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+    actor_dispatch! {
+        Invoke => invoke,
+    }
+}
+```
+
+[fendermint/actors/customsyscall/src/actor.rs#L14](https://github.com/consensus-shipyard/ipc/blob/98497363a10e08236325e6d5c52755b9fcd52958/fendermint/actors/customsyscall/src/actor.rs#L14)
+
+Even though this is Rust code, IPC will compile it as a Wasm target and then run the compiled Wasm code inside FVM as an actor. However, we want to share some of the code between Wasm and IPC, such as the actor name `CUSTOMSYSCALL_ACTOR_NAME` and the `Invoke` method enum. We will define these in a separate file called `[shared.rs](http://shared.rs/)` as follows:
```
Comment (folex) [inline specs/customizability.md:258]: > http://shared.rs/

that's a placeholder?

File: specs/customizability.md [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Invoke = frc42_dispatch::method_hash!("Invoke"),
+}
+```
+
+[fendermint/actors/customsyscall/src/shared.rs](https://github.com/consensus-shipyard/ipc/blob/98497363a10e08236325e6d5c52755b9fcd52958/fendermint/actors/customsyscall/src/shared.rs#L1)
+
+We next need to write a `[lib.rs](http://lib.rs/)` file which exports the shared code and only compiles `actor.rs` if we are building the Wasm actor:
```
Comment (folex) [inline specs/customizability.md:274]: > http://lib.rs/

placeholder?

File: specs/customizability.md (hunk 2 above)
Comment (folex) [inline specs/customizability.md:258]:
```suggestion
Even though this is Rust code, IPC will compile it as a Wasm target and then run the compiled Wasm code inside FVM as an actor. However, we want to share some of the code between Wasm and IPC, such as the actor name `CUSTOMSYSCALL_ACTOR_NAME` and the `Invoke` method enum. We will define these in a separate file called `[shared.rs](https://github.com/consensus-shipyard/ipc/blob/98497363a10e08236325e6d5c52755b9fcd52958/fendermint/actors/customsyscall/src/shared.rs)` as follows:
```

File: specs/customizability.md (hunk 3 above)
Comment (folex) [inline specs/customizability.md:274]:
```suggestion
We next need to write a `[lib.rs](https://github.com/consensus-shipyard/ipc/blob/98497363a10e08236325e6d5c52755b9fcd52958/fendermint/actors/customsyscall/src/lib.rs)` file which exports the shared code and only compiles `actor.rs` if we are building the Wasm actor:
```

File: specs/customizability.md [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+        None,
+    )
+    .context("failed to create customsyscall actor")?;
+```
+
+[fendermint/vm/interpreter/src/fvm/genesis.rs#L251](https://github.com/consensus-shipyard/ipc/blob/98497363a10e08236325e6d5c52755b9fcd52958/fendermint/vm/interpreter/src/fvm/genesis.rs#L251)
+
+Your actor has now been deployed and we should be able to send it messages!
+
+# Genesis Parameters
```
Comment (folex) [inline specs/customizability.md:320]: Should this maybe be a separate spec doc?

File: specs/customizability.md [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+The contents of the `genesis.json` file are essentially the `Genesis` structure defined in the [genesis](https://github.com/consensus-shipyard/ipc/tree/specs/fendermint/vm/genesis) crate. It has the following properties:
+
+- `chain_name` is an arbitrary name for the chain, which will be hashed to become the numeric `chain_id`; in the context of IPC subnets the `chain_name` is expected to be the textual representation of the `SubnetID`
+- `timestamp` is the number of seconds since UNIX epoch
```
Comment (folex) [inline specs/customizability.md:325]:
```suggestion
- `timestamp` is the UNIX timestamp of a moment when genesis was created by the first Validator
```

something like this, maybe?

File: specs/customizability.md [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+- `network_version` is used by the FVM to select gas pricing policy
+- `base_fee` is measured in *atto* and represents the base price for gas
+- `power_scale` is the number of decimals to take into account from the FIL token collateral balance when converting it into voting power expressed as `u64`, which is what CometBFT expects. For example if the scale is 0 then every 1 FIL gives 1 voting power, if it’s 3 then every 0.001 FIL does, and if it’s -1 then every 10 FIL. The power is rounded upwards, so that we don’t end up with 0 power, which would be rejected by CometBFT; for example if the scale is 1, then both 1.1 FIL and 1.9 FIL give 2 power.
```
Comment (folex) [inline specs/customizability.md:328]: maybe mention that it only works for the `Collateral` (ie PoS) mode?

File: specs/customizability.md [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+
+The `genesis.json` file is usually constructed in one of two ways:
+
+- Using the `fendermint genesis` [CLI commands](https://github.com/consensus-shipyard/ipc/blob/specs/fendermint/app/options/src/genesis.rs) to incrementally add validators and accounts as described [here](https://github.com/consensus-shipyard/ipc/blob/specs/docs/fendermint/running.md#genesis).
+- Using the `fenderming genesis ipc from-parent` command to fetch the parameters from the subnet actor on the parent subnet and export them to file.
+
+# Gas policy
+
+The gas price in the FVM is by default [determined](https://github.com/filecoin-project/ref-fvm/blob/c39d880d086aa2e771c7190163436e02715d80f3/fvm/src/machine/mod.rs#L156) by the network version when the `NetworkConfig` is created, but could further be customised by assigning to the `price_list` field. Should Fendermint have to do that, it would be in the [constructor](https://github.com/consensus-shipyard/ipc/blob/7af25c4c860f5ab828e8177927a0f8b6b7a7cc74/fendermint/vm/interpreter/src/fvm/state/exec.rs#L132) of the `FvmExecState`.
```
Comment (folex) [inline specs/customizability.md:346]: > assigning to the `price_list` field

`price_list` is hypothetical here or where does it come from?

File: specs/customizability.md [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+
+The `base_fee` set in genesis is also part of the gas policy. Currently it’s a static value, but it’s part of the upgradable parts of the `FvmExecState` and `FvmStateParams`, with its history maintained by the `App`. The interpreters could have logic to change its value based on the blocks they process.
+
+Also take a look at the following issue to see another proposal to use the `chainmetadata` actor to track the `base_fee` :
```
Comment (folex) [inline specs/customizability.md:350]: Add TODO or link to the issue?

File: specs/customizability.md (hunk 5 above)
Comment (fridrik01) [reply specs/customizability.md:325]: Great, thanks!

File: specs/customizability.md (hunk 8 above)
Comment (fridrik01) [reply specs/customizability.md:350]: The link is to an internal linear issue, will remove this and mention adding this as a GH issue

**Review by raulk — APPROVED** [review]
The first half is duplicative with the relevant section in the developer docs, and isn't really a specification. Will merge for now and clean up the redundancy once we also merge #1014.

File: specs/customizability.md (hunk 7 above)
Comment (raulk) [reply specs/customizability.md:346]: The FVM will return a `price_list` for the specified Filecoin network version, but the user can customize it before creating an FVM Machine.

File: specs/customizability.md (hunk 6 above)
Comment (raulk) [reply specs/customizability.md:328]: So... as we've noted recently, this applies in `Federated` mode too.

File: specs/customizability.md (hunk 4 above)
Comment (raulk) [reply specs/customizability.md:320]: Yeah, definitely belongs in a separate doc.

**Review by raulk — APPROVED** [review]

**Pull request merged**